### PR TITLE
Implement CI in GHA

### PR DESCRIPTION
**Pull request by fjammes**
- Add global configuration file
- Add script to install and configure Spark
- Run minikube in CI
- Run integration tests in CI
- Bump k8s version to 1.25
- Add image analysis in CI
- Update doc

### CODE REVIEW

**Comment by fjammes** [comment]
2 additional features to implements in this PR:

- [x] add secret for KAFKA socket

- [x] move images between CI tasks

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+          image: "${{ env.FINK_K8S_IMAGE }}"
+          fail-build: false
+      - name: Display SARIF report
+        run: |
+          cat ${{ steps.scan.outputs.sarif }}
+      - name: upload Anchore scan SARIF report
+        uses: github/codeql-action/upload-sarif@v2
+        with:
+          sarif_file: ${{ steps.scan.outputs.sarif }}
+  # e2e:
```
Comment (JulienPeloton) [inline .github/workflows/ci.yml]: You can remove commented code below

File: README.md [hunk 2]
```diff
@@ -24,93 +33,79 @@ spark-submit --master $MASTERURL \
...
+## Pre-requisites
+
+Clone the source code:
+
+```bash
+git clone https://github.com/astrolabsoftware/fink-k8s.git
+FINKKUB=$PWD/fink-k8s
+```
+
+All configuration parameters are available in `$PWD/fink-k8s/conf.sh`.
```
Comment (JulienPeloton) [inline README.md]: Better to use the defined variable:

```markdown
All configuration parameters are available in `$FINKKUB/conf.sh`
```

File: docker/Dockerfile_fink [hunk 3]
```diff
@@ -49,6 +49,7 @@ RUN wget --quiet https://repo.anaconda.com/miniconda/Miniconda3-${PYTHON_VERSION
 
 ENV PATH $HOME/miniconda/bin:$PATH
 
+# FIXME clone a fixed version instead of master??
```
Comment (JulienPeloton) [inline docker/Dockerfile_fink:52]: For fink-broker, what we can do is to install the latest tag:

```bash
# in ./bin/install_broker.sh

git clone https://github.com/astrolabsoftware/fink-broker.git
cd fink-broker

# Get new tags from remote
git fetch --tags

# Get latest tag name
latestTag=$(git describe --tags `git rev-list --tags --max-count=1`)

# Checkout latest tag
git checkout $latestTag
```

(probably with a better management for the paths than doing `cd`...). For fink-docker however, let's keep it as is (we will merge it with fink-k8s one day).

File: README.md [hunk 4]
```diff
@@ -24,93 +33,79 @@ spark-submit --master $MASTERURL \
...
 
 ```bash
-cd ${SPARK_HOME}
 
-# remove the `-m` option if you are not using minikube
-# use your docker account in -r if you are not using minikube
-# tag contains fink version and spark version used
-./bin/docker-image-tool-fink.sh -m -r test -t 2.4_3.1.3 -p ./kubernetes/dockerfiles/spark/bindings/python/Dockerfile_fink build
+# Use the `-m` option only if you are using minikube
+$FINKKUB/bin/docker-image-tool-fink.sh -m
```
Comment (JulienPeloton) [inline README.md]: When I execute this command, it just prints the help. It should build the image as well?

```
$FINKKUB/bin/docker-image-tool-fink.sh -m

+ SPARK_ENV_LOADED=
+ SPARK_SCALA_VERSION=
+++ dirname /home/peloton/codes/fink-k8s/bin/docker-image-tool-fink.sh
++ cd /home/peloton/codes/fink-k8s/bin
++ pwd -P
+ readonly DIR=/home/peloton/codes/fink-k8s/bin
+ DIR=/home/peloton/codes/fink-k8s/bin
++ readlink -f /home/peloton/codes/fink-k8s/bin/..
+ readonly FINKKUB=/home/peloton/codes/fink-k8s
+ FINKKUB=/home/peloton/codes/fink-k8s
+ . /home/peloton/codes/fink-k8s/conf.sh
++ K8S_VERSION=v1.25.1
++ CPUS=4
++ readonly SPARK_VERSION=3.1.3
++ SPARK_VERSION=3.1.3
++ readonly SPARK_NAME=spark-3.1.3-bin-hadoop3.2
++ SPARK_NAME=spark-3.1.3-bin-hadoop3.2
++ readonly SPARK_INSTALL_DIR=/home/peloton/fink-k8s-tmp
++ SPARK_INSTALL_DIR=/home/peloton/fink-k8s-tmp
++ export SPARK_HOME=/home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2
++ SPARK_HOME=/home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2
++ export PATH=/home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2/bin:/home/peloton/anaconda3/bin:/home/peloton/anaconda3/condabin:/home/peloton/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin:/home/peloton/codes/fink-broker/bin:/home/peloton/codes/fink-client/fink_client/scripts:/home/peloton/workspace/spark-3.1.3-bin-hadoop3.2/bin
++ PATH=/home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2/bin:/home/peloton/anaconda3/bin:/home/peloton/anaconda3/condabin:/home/peloton/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin:/home/peloton/codes/fink-broker/bin:/home/peloton/codes/fink-client/fink_client/scripts:/home/peloton/workspace/spark-3.1.3-bin-hadoop3.2/bin
++ readonly REPO=local
++ REPO=local
++ readonly TAG=2.4_3.1.3
++ TAG=2.4_3.1.3
++ readonly FINK_IMAGE_NAME=finkk8sdev
++ FINK_IMAGE_NAME=finkk8sdev
++ readonly FINK_K8S_IMAGE=local/finkk8sdev:2.4_3.1.3
++ FINK_K8S_IMAGE=local/finkk8sdev:2.4_3.1.3
++ KAFKA_SOCKET=128.0.0.1:24499
++ KAFKA_TOPIC=changemetopicname
+ . /home/peloton/codes/fink-k8s/bin/include.sh
+ '[' -z /home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2 ']'
+ . /home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2/bin/load-spark-env.sh
++ '[' -z /home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2 ']'
++ SPARK_ENV_SH=spark-env.sh
++ '[' -z '' ']'
++ export SPARK_ENV_LOADED=1
++ SPARK_ENV_LOADED=1
++ export SPARK_CONF_DIR=/home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2/conf
++ SPARK_CONF_DIR=/home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2/conf
++ SPARK_ENV_SH=/home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2/conf/spark-env.sh
++ [[ -f /home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2/conf/spark-env.sh ]]
++ '[' -z '' ']'
++ SCALA_VERSION_1=2.13
++ SCALA_VERSION_2=2.12
++ ASSEMBLY_DIR_1=/home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2/assembly/target/scala-2.13
++ ASSEMBLY_DIR_2=/home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2/assembly/target/scala-2.12
++ ENV_VARIABLE_DOC=https://spark.apache.org/docs/latest/configuration.html#environment-variables
++ [[ -d /home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2/assembly/target/scala-2.13 ]]
++ [[ -d /home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2/assembly/target/scala-2.13 ]]
++ export SPARK_SCALA_VERSION=2.12
++ SPARK_SCALA_VERSION=2.12
+ CTX_DIR=/home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2/target/tmp/docker
+ trap cleanup_ctx_dir EXIT
+ [[ -m = *--help ]]
+ [[ -m = *-h ]]
+ BASEDOCKERFILE=
+ PYDOCKERFILE=
+ RDOCKERFILE=
+ NOCACHEARG=
+ BUILD_PARAMS=
+ SPARK_UID=
+ CROSS_BUILD=false
+ getopts f:p:R:mXnb:u: option
+ case "${option}" in
+ which minikube
+ minikube status
++ minikube docker-env --shell bash
+ eval export 'DOCKER_TLS_VERIFY="1"' export 'DOCKER_HOST="tcp://192.168.49.2:2376"' export 'DOCKER_CERT_PATH="/home/peloton/.minikube/certs"' export 'MINIKUBE_ACTIVE_DOCKERD="minikube"' '#' To point your shell to 'minikube'\''s' docker-daemon, run: '#' eval '$(minikube' -p minikube 'docker-env)'
++ export DOCKER_TLS_VERIFY=1 export DOCKER_HOST=tcp://192.168.49.2:2376 export DOCKER_CERT_PATH=/home/peloton/.minikube/certs export MINIKUBE_ACTIVE_DOCKERD=minikube
++ DOCKER_TLS_VERIFY=1
++ DOCKER_HOST=tcp://192.168.49.2:2376
++ DOCKER_CERT_PATH=/home/peloton/.minikube/certs
++ MINIKUBE_ACTIVE_DOCKERD=minikube
+ getopts f:p:R:mXnb:u: option
+ case "${@: -1}" in
+ usage
+ cat
Usage: /home/peloton/codes/fink-k8s/bin/docker-image-tool-fink.sh [options] [command]
Builds or pushes the built-in Spark Docker image.

Commands:
  build       Build image. Requires a repository address to be provided if the image will be
              pushed to a different registry.
  push        Push a pre-built image to a registry. Requires a repository address to be provided.

Options:
  -f file               Dockerfile to build for JVM based Jobs. By default builds the Dockerfile shipped with Spark.
  -p file               (Optional) Dockerfile to build for PySpark Jobs. Builds Python dependencies and ships with Spark.
                        Skips building PySpark docker image if not specified.
  -R file               (Optional) Dockerfile to build for SparkR Jobs. Builds R dependencies and ships with Spark.
                        Skips building SparkR docker image if not specified.
  -m                    Use minikube's Docker daemon.
  -n                    Build docker image with --no-cache
  -u uid                UID to use in the USER directive to set the user the main Spark process runs as inside the
                        resulting container
  -X                    Use docker buildx to cross build. Automatically pushes.
                        See https://docs.docker.com/buildx/working-with-buildx/ for steps to setup buildx.
  -b arg                Build arg to build or push the image. For multiple build args, this option needs to
                        be used separately for each build arg.

Using minikube when building images will do so directly into minikube's Docker daemon.
There is no need to push the images into minikube in that case, they'll be automatically
available when running applications inside the minikube cluster.

Check the following documentation for more information on using the minikube Docker daemon:

  https://kubernetes.io/docs/getting-started-guides/minikube/#reusing-the-docker-daemon

Examples:
  - Build image in minikube with tag "testing"
    /home/peloton/codes/fink-k8s/bin/docker-image-tool-fink.sh -m -t testing build

  - Build PySpark docker image
    /home/peloton/codes/fink-k8s/bin/docker-image-tool-fink.sh -r docker.io/myrepo -t v2.3.0 -p kubernetes/dockerfiles/spark/bindings/python/Dockerfile build

  - Build and push image with tag "v2.3.0" to docker.io/myrepo
    /home/peloton/codes/fink-k8s/bin/docker-image-tool-fink.sh -r docker.io/myrepo -t v2.3.0 build
    /home/peloton/codes/fink-k8s/bin/docker-image-tool-fink.sh -r docker.io/myrepo -t v2.3.0 push

  - Build and push JDK11-based image with tag "v3.0.0" to docker.io/myrepo
    /home/peloton/codes/fink-k8s/bin/docker-image-tool-fink.sh -r docker.io/myrepo -t v3.0.0 -b java_image_tag=11-jre-slim build
    /home/peloton/codes/fink-k8s/bin/docker-image-tool-fink.sh -r docker.io/myrepo -t v3.0.0 push

  - Build and push JDK11-based image for multiple archs to docker.io/myrepo
    /home/peloton/codes/fink-k8s/bin/docker-image-tool-fink.sh -r docker.io/myrepo -t v3.0.0 -X -b java_image_tag=11-jre-slim build
    # Note: buildx, which does cross building, needs to do the push during build
    # So there is no separate push step with -X

+ exit 1
+ cleanup_ctx_dir
+ is_dev_build
+ '[' '!' -f /home/peloton/fink-k8s-tmp/spark-3.1.3-bin-hadoop3.2/RELEASE ']'
```

**Review by JulienPeloton — COMMENTED** [review]
Thanks @fjammes ! I followed the instructions, but I got stuck when launching `$FINKKUB/bin/docker-image-tool-fink.sh -m`. I also left some minor in-lined comments.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (fjammes) [reply .github/workflows/ci.yml]: :+1:

File: README.md (hunk 2 above)
Comment (fjammes) [reply README.md]: My mistake, you need to run `$FINKKUB/bin/docker-image-tool-fink.sh -m build`. I fixed the doc.

Comment (fjammes) [reply README.md]: :+1:

File: docker/Dockerfile_fink (hunk 3 above)
Comment (fjammes) [reply docker/Dockerfile_fink:52]: This look safer to me:
```
ARG BROKER_VERSION 2.5
RUN git clone --depth 1 -b "$BROKER_VERSION" --single-branch https://github.com/astrolabsoftware/fink-broker.git
```
Then we need to explicitly move the fink-broker version, for example when packaging a new release.

Comment (fjammes) [reply docker/Dockerfile_fink:52]: Would it work for you?

File: README.md (hunk 4 above)
Comment (fjammes) [reply README.md]: The `build` argument was missing in the doc, in addition, I removed the debug flag

**Pull request merged**